File: vpp-tools/vpp_wireguard.py
#!/usr/bin/env python3
"""VectorOS VPP WireGuard Manager - Direct VPP WireGuard plugin management"""

import sys
import json
import subprocess
import base64
import os

def run_vppctl(cmd):
    """Run vppctl command"""
    try:
        result = subprocess.run(
            ['vppctl'] + cmd.split(),
            capture_output=True,
            text=True,
            timeout=10
        )
        return result.stdout.strip(), result.stderr.strip(), result.returncode
    except Exception as e:
        return '', str(e), 1

def generate_keypair():
    """Generate WireGuard keypair"""
    try:
        # Generate private key
        result = subprocess.run(['wg', 'genkey'], capture_output=True, text=True)
        if result.returncode != 0:
            return {'error': 'Failed to generate private key'}
        private_key = result.stdout.strip()

        # Generate public key from private key
        result = subprocess.run(['wg', 'pubkey'], input=private_key, capture_output=True, text=True)
        if result.returncode != 0:
            return {'error': 'Failed to generate public key'}
        public_key = result.stdout.strip()

        return {
            'private_key': private_key,
            'public_key': public_key
        }
    except FileNotFoundError:
        return {'error': 'wg tool not installed'}

def create_interface(listen_port, private_key, src_ip=None):
    """Create WireGuard interface in VPP"""
    cmd = f'wireguard create listen-port {listen_port} private-key {private_key}'
    if src_ip:
        cmd += f' src {src_ip}'

    stdout, stderr, rc = run_vppctl(cmd)
    if rc != 0:
        return {'error': stderr or 'Failed to create WireGuard interface'}

    # Parse interface name from output
    iface_name = stdout.strip() if stdout else 'wg0'
    return {
        'status': 'ok',
        'interface': iface_name,
        'message': f'WireGuard interface {iface_name} created'
    }

def delete_interface(iface_name):
    """Delete WireGuard interface"""
    stdout, stderr, rc = run_vppctl(f'wireguard {iface_name} delete')
    if rc != 0:
        return {'error': stderr or 'Failed to delete WireGuard interface'}
    return {'status': 'ok', 'message': f'WireGuard interface {iface_name} deleted'}

def add_peer(iface_name, public_key, endpoint=None, port=None, allowed_ips=None, preshared_key=None):
    """Add peer to WireGuard interface"""
    cmd = f'wireguard peer add {iface_name} public-key {public_key}'

    if endpoint and port:
        cmd += f' endpoint {endpoint} port {port}'

    if allowed_ips:
        for ip in allowed_ips:
            cmd += f' allowed-ip {ip}'

    if preshared_key:
        cmd += f' preshared-key {preshared_key}'

    stdout, stderr, rc = run_vppctl(cmd)
    if rc != 0:
        return {'error': stderr or 'Failed to add peer'}

    return {'status': 'ok', 'message': 'Peer added successfully'}

def remove_peer(iface_name, public_key):
    """Remove peer from WireGuard interface"""
    stdout, stderr, rc = run_vppctl(f'wireguard peer remove {iface_name} public-key {public_key}')
    if rc != 0:
        return {'error': stderr or 'Failed to remove peer'}
    return {'status': 'ok', 'message': 'Peer removed successfully'}

def show_interfaces():
    """Show WireGuard interfaces"""
    stdout, stderr, rc = run_vppctl('show wireguard interface')
    if rc != 0:
        return {'error': stderr or 'Failed to show interfaces'}

    interfaces = []
    for line in stdout.split('\n'):
        line = line.strip()
        if line and not line.startswith('Interface') and not line.startswith('---'):
            parts = line.split()
            if len(parts) >= 2:
                interfaces.append({
                    'name': parts[0],
                    'index': parts[1] if len(parts) > 1 else '',
                    'state': parts[2] if len(parts) > 2 else 'unknown'
                })

    return {'interfaces': interfaces}

def show_peers():
    """Show WireGuard peers"""
    stdout, stderr, rc = run_vppctl('show wireguard peer')
    if rc != 0:
        return {'error': stderr or 'Failed to show peers'}

    peers = []
    for line in stdout.split('\n'):
        line = line.strip()
        if line and not line.startswith('Peer') and not line.startswith('---'):
            parts = line.split()
            if len(parts) >= 2:
                peers.append({
                    'interface': parts[0] if len(parts) > 0 else '',
                    'public_key': parts[1] if len(parts) > 1 else '',
                    'endpoint': parts[2] if len(parts) > 2 and parts[2] != '(none)' else None,
                    'allowed_ips': parts[3:] if len(parts) > 3 else []
                })

    return {'peers': peers}

def main():
    import argparse
    parser = argparse.ArgumentParser(description='VectorOS VPP WireGuard Manager')
    parser.add_argument('action', choices=['create', 'delete', 'peer-add', 'peer-remove', 'show', 'peers', 'genkey'])
    parser.add_argument('--interface', help='WireGuard interface name')
    parser.add_argument('--listen-port', type=int, default=51820)
    parser.add_argument('--private-key', help='Private key')
    parser.add_argument('--src-ip', help='Source IP address')
    parser.add_argument('--public-key', help='Peer public key')
    parser.add_argument('--endpoint', help='Peer endpoint')
    parser.add_argument('--port', type=int, help='Peer port')
    parser.add_argument('--allowed-ips', nargs='+', help='Allowed IPs')
    parser.add_argument('--preshared-key', help='Preshared key')

    args = parser.parse_args()

    try:
        if args.action == 'genkey':
            result = generate_keypair()
        elif args.action == 'create':
            result = create_interface(args.listen_port, args.private_key, args.src_ip)
        elif args.action == 'delete':
            result = delete_interface(args.interface)
        elif args.action == 'peer-add':
            result = add_peer(args.interface, args.public_key, args.endpoint, args.port, args.allowed_ips, args.preshared_key)
        elif args.action == 'peer-remove':
            result = remove_peer(args.interface, args.public_key)
        elif args.action == 'show':
            result = show_interfaces()
        elif args.action == 'peers':
            result = show_peers()

        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({'error': str(e)}))
        sys.exit(1)

if __name__ == '__main__':
    main()

File: control-plane/src/services/vpp_wireguard.rs
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::process::Command;
use tracing::info;

/// VPP WireGuard Manager
pub struct VppWireGuardManager {
    script_path: String,
}

impl VppWireGuardManager {
    pub fn new() -> Self {
        Self {
            script_path: "/root/VectorOS/vpp-tools/vpp_wireguard.py".to_string(),
        }
    }

    /// Generate WireGuard keypair
    pub fn generate_keypair(&self) -> Result<(String, String)> {
        let output = Command::new("python3")
            .arg(&self.script_path)
            .arg("genkey")
            .output()?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("Failed to generate keypair: {}", stderr);
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let result: serde_json::Value = serde_json::from_str(&stdout)?;

        if let Some(error) = result.get("error") {
            anyhow::bail!("Keypair error: {}", error);
        }

        let private_key = result.get("private_key")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let public_key = result.get("public_key")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        Ok((private_key, public_key))
    }

    /// Create WireGuard interface
    pub fn create_interface(&self, listen_port: u16, private_key: &str, src_ip: Option<&str>) -> Result<String> {
        let mut args = vec![
            "create".to_string(),
            "--listen-port".to_string(),
            listen_port.to_string(),
            "--private-key".to_string(),
            private_key.to_string(),
        ];

        if let Some(ip) = src_ip {
            args.push("--src-ip".to_string());
            args.push(ip.to_string());
        }

        let output = Command::new("python3")
            .arg(&self.script_path)
            .args(&args)
            .output()?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("Failed to create WireGuard interface: {}", stderr);
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let result: serde_json::Value = serde_json::from_str(&stdout)?;

        if let Some(error) = result.get("error") {
            anyhow::bail!("Create interface error: {}", error);
        }

        Ok(result.get("interface")
            .and_then(|v| v.as_str())
            .unwrap_or("wg0")
            .to_string())
    }

    /// Add peer to WireGuard interface
    pub fn add_peer(
        &self,
        iface_name: &str,
        public_key: &str,
        endpoint: Option<&str>,
        port: Option<u16>,
        allowed_ips: &[String],
        preshared_key: Option<&str>,
    ) -> Result<()> {
        let mut args = vec![
            "peer-add".to_string(),
            "--interface".to_string(),
            iface_name.to_string(),
            "--public-key".to_string(),
            public_key.to_string(),
        ];

        if let Some(ep) = endpoint {
            args.push("--endpoint".to_string());
            args.push(ep.to_string());
        }

        if let Some(p) = port {
            args.push("--port".to_string());
            args.push(p.to_string());
        }

        if !allowed_ips.is_empty() {
            args.push("--allowed-ips".to_string());
            args.extend(allowed_ips.iter().cloned());
        }

        if let Some(psk) = preshared_key {
            args.push("--preshared-key".to_string());
            args.push(psk.to_string());
        }

        let output = Command::new("python3")
            .arg(&self.script_path)
            .args(&args)
            .output()?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("Failed to add peer: {}", stderr);
        }

        Ok(())
    }

    /// Show WireGuard interfaces
    pub fn show_interfaces(&self) -> Result<Vec<WireGuardInterface>> {
        let output = Command::new("python3")
            .arg(&self.script_path)
            .arg("show")
            .output()?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("Failed to show interfaces: {}", stderr);
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let result: serde_json::Value = serde_json::from_str(&stdout)?;

        let mut interfaces = Vec::new();
        if let Some(arr) = result.get("interfaces").and_then(|v| v.as_array()) {
            for iface in arr {
                interfaces.push(WireGuardInterface {
                    name: iface.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                    index: iface.get("index").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                    state: iface.get("state").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                });
            }
        }

        Ok(interfaces)
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WireGuardInterface {
    pub name: String,
    pub index: String,
    pub state: String,
}
